### PR TITLE
Test create initial category

### LINKED ISSUE #847: Creating a category is impossible in the fresh installation with category table empty.
#### Sonata packages

<details><summary>show</summary>
<p>

```
sonata-project/admin-bundle              4.12.0            4.12.0          The missing Symfony Admin Generator
sonata-project/block-bundle              4.13.0            4.13.0          Symfony SonataBlockBundle
sonata-project/cache                     2.2.0             2.2.0           Cache library
Package sonata-project/cache is abandoned, you should avoid using it. No replacement was suggested.
sonata-project/classification-bundle     4.0.4             4.0.4           Symfony SonataClassificationBundle
sonata-project/doctrine-extensions       1.17.0            1.17.0          Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 4.3.0             4.3.0           Integrate Doctrine ORM into the SonataAd...
sonata-project/exporter                  2.13.0            2.13.0          Lightweight Exporter library
sonata-project/form-extensions           1.17.0            1.17.0          Symfony form extensions
sonata-project/formatter-bundle          5.0.x-dev 5ecbc1c 5.x-dev 6bc6e6a Symfony SonataFormatterBundle
sonata-project/media-bundle              4.2.0             4.2.0           Symfony SonataMediaBundle
sonata-project/twig-extensions           1.10.0            1.10.0          Sonata twig extensions
sonata-project/user-bundle               5.1.0             5.1.0           Symfony SonataUserBundle
```

</p>
</details>

#### Symfony packages

<details><summary>show</summary>
<p>

```
symfony/asset                      v6.1.0  v6.1.0  Manages URL generation and versioning of web assets such as CSS ...
symfony/cache                      v6.1.1  v6.1.1  Provides an extended PSR-6, PSR-16 (and tags) implementation
symfony/cache-contracts            v3.1.0  v3.1.0  Generic abstractions related to caching
symfony/config                     v6.1.0  v6.1.0  Helps you find, load, combine, autofill and validate configurati...
symfony/console                    v6.1.1  v6.1.1  Eases the creation of beautiful and testable command line interf...
symfony/dependency-injection       v6.1.0  v6.1.0  Allows you to standardize and centralize the way objects are con...
symfony/deprecation-contracts      v3.1.0  v3.1.0  A generic function and convention to trigger deprecation notices
symfony/doctrine-bridge            v6.1.0  v6.1.0  Provides integration for Doctrine with various Symfony components
symfony/dotenv                     v6.1.0  v6.1.0  Registers environment variables from a .env file
symfony/error-handler              v6.1.0  v6.1.0  Provides tools to manage errors and ease debugging PHP code
symfony/event-dispatcher           v6.1.0  v6.1.0  Provides tools that allow your application components to communi...
symfony/event-dispatcher-contracts v3.1.0  v3.1.0  Generic abstractions related to dispatching event
symfony/expression-language        v6.1.0  v6.1.0  Provides an engine that can compile and evaluate expressions
symfony/filesystem                 v6.1.0  v6.1.0  Provides basic utilities for the filesystem
symfony/finder                     v6.1.0  v6.1.0  Finds files and directories via an intuitive fluent interface
symfony/flex                       v2.2.2  v2.2.2  Composer plugin for Symfony
symfony/form                       v6.1.1  v6.1.1  Allows to easily create, process and reuse HTML forms
symfony/framework-bundle           v6.1.1  v6.1.1  Provides a tight integration between Symfony components and the ...
symfony/http-client                v6.1.1  v6.1.1  Provides powerful methods to fetch HTTP resources synchronously ...
symfony/http-client-contracts      v3.1.0  v3.1.0  Generic abstractions related to HTTP clients
symfony/http-foundation            v6.1.1  v6.1.1  Defines an object-oriented layer for the HTTP specification
symfony/http-kernel                v6.1.1  v6.1.1  Provides a structured process for converting a Request into a Re...
symfony/mailer                     v6.1.1  v6.1.1  Helps sending emails
symfony/maker-bundle               v1.43.0 v1.43.0 Symfony Maker helps you create empty commands, controllers, form...
symfony/mime                       v6.1.1  v6.1.1  Allows manipulating MIME messages
symfony/options-resolver           v6.1.0  v6.1.0  Provides an improved replacement for the array_replace PHP function
symfony/password-hasher            v6.1.0  v6.1.0  Provides password hashing utilities
symfony/polyfill-intl-grapheme     v1.26.0 v1.26.0 Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-icu          v1.26.0 v1.26.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn          v1.26.0 v1.26.0 Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer   v1.26.0 v1.26.0 Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring          v1.26.0 v1.26.0 Symfony polyfill for the Mbstring extension
symfony/property-access            v6.1.0  v6.1.0  Provides functions to read and write from/to an object or array ...
symfony/property-info              v6.1.1  v6.1.1  Extracts information about PHP class' properties using metadata ...
symfony/proxy-manager-bridge       v6.1.0  v6.1.0  Provides integration for ProxyManager with various Symfony compo...
symfony/routing                    v6.1.1  v6.1.1  Maps an HTTP request to a set of configuration variables
symfony/runtime                    v6.1.1  v6.1.1  Enables decoupling PHP applications from global state
symfony/security-acl               v3.3.1  v3.3.1  Symfony Security Component - ACL (Access Control List)
symfony/security-bundle            v6.1.0  v6.1.0  Provides a tight integration of the Security component into the ...
symfony/security-core              v6.1.0  v6.1.0  Symfony Security Component - Core Library
symfony/security-csrf              v6.1.0  v6.1.0  Symfony Security Component - CSRF Library
symfony/security-http              v6.1.1  v6.1.1  Symfony Security Component - HTTP Integration
symfony/serializer                 v6.1.1  v6.1.1  Handles serializing and deserializing data structures, including...
symfony/service-contracts          v3.1.0  v3.1.0  Generic abstractions related to writing services
symfony/stopwatch                  v6.1.0  v6.1.0  Provides a way to profile code
symfony/string                     v6.1.0  v6.1.0  Provides an object-oriented API to strings and deals with bytes,...
symfony/translation                v6.1.0  v6.1.0  Provides tools to internationalize your application
symfony/translation-contracts      v3.1.0  v3.1.0  Generic abstractions related to translation
symfony/twig-bridge                v6.1.0  v6.1.0  Provides integration for Twig with various Symfony components
symfony/twig-bundle                v6.1.1  v6.1.1  Provides a tight integration of Twig into the Symfony full-stack...
symfony/validator                  v6.1.1  v6.1.1  Provides tools to validate values
symfony/var-dumper                 v6.1.0  v6.1.0  Provides mechanisms for walking through any arbitrary PHP variable
symfony/var-exporter               v6.1.1  v6.1.1  Allows exporting any serializable PHP data structure to plain PH...
symfony/web-link                   v6.1.0  v6.1.0  Manages links between resources
symfony/web-profiler-bundle        v6.1.1  v6.1.1  Provides a development tool that gives detailed information abou...
symfony/yaml                       v6.1.0  v6.1.0  Loads and dumps YAML files
```

</p>
</details>

#### PHP version

```
PHP 8.1.5 (cli) (built: Apr 12 2022 17:38:30) (NTS Visual C++ 2019 x64)
```

## Subject

The parent field on create category form is required. When no parent category is selected notification:
This value should not be null.
appears.
Seems related to this frament 
        if ($this->hasSubject()) {
            if (null !== $this->getSubject()->getParent() || null === $this->getSubject()->getId()) { // root category cannot have a parent
                $form
                    ->add('parent', CategorySelectorType::class, [
                        'category' => $this->getSubject(),
                        'model_manager' => $this->getModelManager(),
                        'class' => $this->getClass(),
                        'required' => true,
                        'context' => $this->getSubject()->getContext(),
                    ]);
            }
        }
        
$form ->add('parent', CategorySelectorType::class, [
is executed when the category table is empty - effectively making creation of any category impossible.

## Steps to reproduce
Install bundle with php8 attributes used in entities (not tested if it fails also when annotations are used).
doctrine:schema:update --force
Go to categories in admin, add a category.
Dropdown Parent is marked required and has no values.
Write name.
Submit.

## Expected results
Category is created
## Actual results
It fails with message:
This value should not be null.

### PR DESCRIPTION
**Pull request by jordisala1991**
## Subject

I am targeting this branch, because this is BC.

Closes #847.

## Changelog

```markdown
### Fixed
- Create a category after a fresh install initializes the first context and the root category.
```

### CODE REVIEW

File: src/Entity/CategoryManager.php [hunk 1]
```diff
@@ -103,6 +103,10 @@ public function getAllRootCategories(bool $loadChildren = true): array
             ->getQuery()
             ->getResult();
 
+        if ($rootCategories === []) {
```
Comment (jordisala1991) [inline src/Entity/CategoryManager.php]:
```suggestion
        if ([] === $rootCategories) {
```

File: src/Document/CategoryManager.php [hunk 2]
```diff
@@ -102,6 +102,10 @@ public function getAllRootCategories(bool $loadChildren = true): array
             ->execute();
         \assert(\is_array($rootCategories));
 
+        if ($rootCategories === []) {
```
Comment (jordisala1991) [inline src/Document/CategoryManager.php]:
```suggestion
        if ([] === $rootCategories) {
```

File: tests/Functional/Admin/CategoryAdminTest.php [hunk 3]
```diff
@@ -92,6 +92,26 @@ public static function provideFormUrlsCases(): iterable
...
+            'uniqid' => 'category',
+        ]);
+        $client->submitForm('btn_create_and_list', [
+            'category[name]' => 'Name',
+        ]);
+        $client->followRedirect();
+
+        self::assertResponseIsSuccessful();
+
+        $client->request('GET', '/admin/tests/app/category/tree'));
```
Comment (jordisala1991) [inline tests/Functional/Admin/CategoryAdminTest.php]:
```suggestion
        $client->request('GET', '/admin/tests/app/category/tree');
```

**Comment by jordisala1991** [comment]
To be merged with squash, sorry 😅 .

@VincentLanglet , to me the issue is that the method that gathers all root categories does not work like the other methods in the CategoryManager. The others normally try to see if the context is created and if not, it creates the context on the fly. Same for getCategories given a context.

I have added a condition on the get all root categories. If no category is returned from the db (fresh install), the default context is created (if not exist) and the root category for that context too.

That makes the new test, that starts with an empty db pass (it creates two categories in the process, one when you access to the list, and the other when you save).

**Review by VincentLanglet — APPROVED** [review]

**Pull request merged**